synse/websocket.go

```diff
@@ -0,0 +1,515 @@
+		}, nil
+	}
+
+	// Only return a TLS client if its config option is enable.
+	cert, err := setTLS(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &websocket.Dialer{
+		HandshakeTimeout: opts.WebSocket.HandshakeTimeout,
+		TLSClientConfig: &tls.Config{
+			Certificates: []tls.Certificate{cert},
+		},
+	}, nil
+}
+
+// Open opens the websocket connection between the client and Synse Server.
+func (c *websocketClient) Open() error {
+	conn, _, err := c.client.Dial(buildURL(c.scheme, c.options.Address, c.apiVersion, c.entryRoute), nil)
+	if err != nil {
+		return errors.Wrap(err, "failed to open the websocket connection")
+	}
+
+	c.connection = conn
+	return nil
+}
+
+// Close closes the websocket connection between the client and Synse Server.
+// It's up to the user to close the connection after finish using it.
+// FIXME - it makes sense to use defer function with Close. However, since
+// Close returns an error, do user need to check it manually? If yes, how?
+func (c *websocketClient) Close() error {
+	err := c.connection.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
+	if err != nil {
+		return errors.Wrap(err, "failed to close the connection gracefully")
+	}
+
+	return nil
+}
+
+// Status returns the status info. This is used to check if the server
+// is responsive and reachable.
+// NOTE - this method is not applicable for websocket client.
+func (c *websocketClient) Status() (*scheme.Status, error) {
+	return nil, nil
+}
+
+// Version returns the version info.
+func (c *websocketClient) Version() (*scheme.Version, error) {
+	req := scheme.RequestVersion{
+		EventMeta: scheme.EventMeta{
+			ID:    addCounter(),
+			Event: requestVersion,
+		},
+	}
+
+	resp := new(scheme.ResponseVersion)
+	err := c.makeRequest(req, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	// FIXME - should we return the everything from ResponseVersion
+	// or it is fine to return the Data value only since these metadata
+	// such as request id and request won't be much helpful for consumer
+	// anyway? Should the response from websocket client be similar to http
+	// client?
+	return &resp.Data, nil
+}
+
+// Config returns the unified configuration info.
+func (c *websocketClient) Config() (*scheme.Config, error) {
+	req := scheme.RequestConfig{
+		EventMeta: scheme.EventMeta{
+			ID:    addCounter(),
+			Event: requestConfig,
+		},
+	}
+
+	resp := new(scheme.ResponseConfig)
+	err := c.makeRequest(req, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resp.Data, nil
+}
+
+// Plugins returns the summary of all plugins currently registered with
+// Synse Server.
+// NOTE - this method is not applicable for websocket client.
+func (c *websocketClient) Plugins() (*[]scheme.PluginMeta, error) {
```

Comment:
this should be supported. looking back at the spec (https://github.com/vapor-ware/synse-server/blob/v3/staging/proposals/v3/api-websocket.md#plugins) I think it wasn't totally clear, but if no plugin is specified in the request, this should return info on all the plugins